src/cpp/src/speculative_decoding/speculative_decoding_stateful_eagle3.cpp

```diff
@@ -0,0 +1,1242 @@
+        m_sd_metrics.update_draft_accepted_tokens(0, total_draft_accepted);
+        m_sd_metrics.update_draft_generated_len(0, total_draft_generated);
+        m_sd_metrics.update_generated_len(generated_tokens);
+    }
+
+    m_sd_perf_metrics.evaluate_statistics(generate_timer.get_start_time());
+    results.perf_metrics = m_sd_perf_metrics;
+    results.extended_perf_metrics = std::make_shared<SDPerModelsPerfMetrics>(m_sd_perf_metrics);
+
+    log_debug("=== GENERATION END ===");
+    return results;
+}
+
+StatefulEagle3LLMPipeline::SpeculativeResult StatefulEagle3LLMPipeline::run_speculative_iteration(
+    size_t input_token_count,
+    int64_t eos_token_id) {
+    SpeculativeResult result;
+
+    // TODO(top-k): For top-k speculation, this function will need to:
+    // 1. Generate k draft candidates per position (tree structure)
+    // 2. Build tree attention mask for parallel verification
+    // 3. Accept/reject based on tree traversal
+    OPENVINO_ASSERT(m_main_model->get_running_sequence_count() == 1 && m_draft_model->get_running_sequence_count() == 1,
+                    "Eagle3 speculative iteration requires single sequence per model");
+
+    auto target_hidden_states = m_main_model->get_current_sequence()->get_hidden_state();
+    if (!target_hidden_states || target_hidden_states.get_size() == 0) {
+        return result;
+    }
+
+    log_debug("--- Draft Phase Start ---");
+    log_debug("Input: input_token_count=" + std::to_string(input_token_count) +
+              ", hidden_shape=" + format_shape(target_hidden_states.get_shape()));
+
+    // Record pre-draft sequence lengths for potential rollback
+    const size_t pre_draft_draft_len = m_draft_model->get_sequence_length();
+
+    // Step 1: Generate first draft token using target hidden states
+    auto first_result = m_draft_model->forward({.input_token_count = input_token_count,
+                                                .use_target_hidden = true,
+                                                .source_sequence = m_main_model->get_current_sequence()});
+
+    int64_t first_draft_token = std::get<int64_t>(first_result.sampled_tokens);
+    log_debug("First draft token: " + std::to_string(first_draft_token));
+
+    // Collect draft candidates
+    std::vector<int64_t> draft_candidates;
+    draft_candidates.reserve(m_draft_iterations);
+    draft_candidates.push_back(first_draft_token);
+
+    // Append first token to main model (draft model already has it from sampler)
+    m_main_model->append_tokens({first_draft_token});
+
+    // Step 2: Generate additional draft tokens using internal hidden states
+    for (size_t i = 1; i < m_draft_iterations; ++i) {
+        auto more_result = m_draft_model->forward({.input_token_count = 1, .use_target_hidden = false});
+
+        int64_t draft_token = std::get<int64_t>(more_result.sampled_tokens);
+        draft_candidates.push_back(draft_token);
+        m_main_model->append_tokens({draft_token});
```

Comment:
I would better do `append_tokens(...)` in the loop separately, below the given code, just for the clarity that here we are setting inputs for the target model. But feel free to ignore this suggestion, might be just more comments about tokens appending will solve the confusion!